### PR TITLE
ElectroHydraulicPTO Motor Mode

### PR DESCRIPTION
**Pull request by andermi**
WIP

- make solver more robust
- possibly combine rpm, pressure with electronics in solver
- try computing Jacobian explicitly to use in solver
- try smoothing interpolation with higher order splines
- incorporate motor mode to drive piston with user commanded winding current and bias current
- possibly add better battery model (or make a new branch)

@hamilton8415 I resurrected motor mode and made this draft PR so we could discuss things as we work out of this same branch

### CODE REVIEW

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 1]
```diff
@@ -82,14 +82,43 @@ struct ElectroHydraulicSoln : Functor<double>
...
+    const double I = this->I_Wind(x[0U]);
+    const double J_I = this->I_Wind.df(x[0U]);
+    const double T = 1.375 * this->I_Wind.TorqueConstantInLbPerAmp * I;
+    const double dTdx0 = 1.375 * this->I_Wind.TorqueConstantInLbPerAmp * J_I;
+
+    fjac(0, 0) = 1 - J_eff_v[0U] * SecondsPerMinute * this->Q / this->HydMotorDisp;  // df0/dx0
+    fjac(0, 1) = -J_eff_v[1U] * SecondsPerMinute * this->Q / this->HydMotorDisp;  // df0/dx1
+    fjac(1, 0) =
+      (-J_eff_m[0U] * T - eff_m * dTdx0) / (this->HydMotorDisp / (2.0 * M_PI));  // df1/dx0
+    fjac(1, 1) = 1 - J_eff_m[1U] * T / (this->HydMotorDisp / (2.0 * M_PI));  // df1/dx1
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:116]: check my math?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 2]
```diff
@@ -82,14 +82,43 @@ struct ElectroHydraulicSoln : Functor<double>
...
+  static const std::vector<std::vector<double>> Eff_M;  // mechanical efficiency
 
 public:
   ElectroHydraulicSoln()
   : Functor<double>(2, 2),
     // Set HydraulicMotor Volumetric & Mechanical Efficiency
-    hyd_eff_v(Neff, Peff, eff_v), hyd_eff_m(Neff, Peff, eff_m) {}
+    hyd_eff_v(Neff, Peff, Eff_V), hyd_eff_m(Neff, Peff, Eff_M) {}
+
+  int df(const Eigen::VectorXd & x, Eigen::MatrixXd & fjac) const
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:94]: the solver looks for this function in the functor

File: buoy_gazebo/src/ElectroHydraulicPTO/WindingCurrentTarget.hpp [hunk 3]
```diff
@@ -80,16 +81,34 @@ class WindingCurrentTarget
     this->RamPosition = 0;  // Default to full retract, should be set before () operator is used.
   }
 
+  double df(const double & N) const
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/WindingCurrentTarget.hpp:84]: just following the functor naming convention although the solver doesn't call this directly

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 4]
```diff
@@ -115,9 +144,15 @@ struct ElectroHydraulicSoln : Functor<double>
         CubicInchesPerGallon / SecondsPerMinute;
     }
 
-    fvec[0U] = x[0U] - eff_v * SecondsPerMinute * QQ / this->HydMotorDisp;
+    // TODO(anyone) put this back after evaluating simplified solver
+    // fvec[0U] = x[0U] - eff_v * SecondsPerMinute * QQ / this->HydMotorDisp;
+    fvec[0U] = x[0U] - eff_v * SecondsPerMinute * this->Q / this->HydMotorDisp;
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:149]: commented out the relief valve to simplify the problem and evaluate the solver

**Comment by andermi** [comment]
https://github.com/osrf/buoy_sim/blob/decd67c5417d648c7eff11a0004a017ac2df5895/buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp#L212

I didn't put this back in yet just to keep the solver simple for now

**Comment by andermi** [comment]
check the comparison to the test machine:
`ros2 launch buoy_tests experiment_comparison_select.launch.py manual:=true`

**Comment by hamilton8415** [comment]
Very good, thanks for resurrecting this branch, I was looking into how to bring it up to date and you beat me to it (big surprise).  One item perhaps to add to the todo list is a battery model improvement, which may or may not get itself tangled up in the solver.  Currently the model used uses a constant EMF value, and an improvement would be a coloumb counter that would drive this Vemf  up and down as the battery charges and discharges.  The internal resistance of the battery may also be dependent upon the state of charge.  I think this is orthogonal to the other work so maybe should be done in a second branch after this is done if it's not related to the solver, we will see.  In any event, not a required  simulation feature, but would be a nice touch. :)

**Comment by andermi** [comment]
I pushed a commit replacing splinter with the custom interp1d

**Comment by andermi** [comment]
@hamilton8415 just pushed updates to new tests and new launch files for manual vs ci

**Comment by hamilton8415** [comment]
@andermi @mjcarroll:  I completed the linter work and checked the functionality of motormode in terms of the entire model and simulation, and of course also the tests related to the electro-hydraulic PTO.  There are some CI errors, some output below.  I'm hoping one of you could look into what's going in there as it's a bit out of my depth.  I believe andermi did these originally.  Thanks!  Looking forward to getting this off the plate and onward with some smaller pull-requests. :)


2022-11-15T03:22:39.4222419Z The following tests FAILED:
2022-11-15T03:22:39.4224100Z 	 30 - launch_sc_commands_ros_feedback.launch.py (Failed)
2022-11-15T03:22:39.4225675Z 	 34 - launch_sc_pump_ros_feedback_py.launch.py (Failed)
2022-11-15T03:22:39.5858009Z Errors while running CTest
2022-11-15T03:22:39.5889542Z --- stderr: buoy_tests
2022-11-15T03:22:39.5891090Z Errors while running CTest

**Comment by andermi** [comment]
Somehow I guess I did miss your comment here! Looking into it now.

**Comment by andermi** [comment]
I think we are good to go after a review

File: .gitignore [hunk 5]
```diff
@@ -3,3 +3,11 @@
...
 
 __pycache__
+
+splinter/include
+splinter/lib
+splinter/splinter
+splinter/assets
+splinter/CREDITS.md
+splinter/docs
+splinter/README_SPLINTER.md
```
Comment (andermi) [inline .gitignore:13]: could delete splinter reference here

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp [hunk 6]
```diff
@@ -206,10 +212,10 @@ void ElectroHydraulicPTO::PreUpdate(
   // TODO(anyone): Figure out if (0) for the index is always correct,
   // some OR code has a process of finding the index for this argument.
   double xdot = prismaticJointVelComp->Data().at(0);
-  this->dataPtr->functor.Q = xdot * 39.4 * this->dataPtr->PistonArea;  // inch^3/second
+  this->dataPtr->functor.Q = xdot * 39.4 * this->dataPtr->PistonArea;       // inch^3/second
 
-  double x = prismaticJointVelComp->Data().at(0);
-  this->dataPtr->functor.I_Wind.RamPosition = 40.0;  // TODO(hamilton8415) x * 39.4;
+  double PistonPos = prismaticJointVelComp->Data().at(0);
+  this->dataPtr->functor.I_Wind.RamPosition = 40;       // 2.03 - PistonPos * 39.4;
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp:218]: @hamilton8415, with this like it is, I don't think it's using your ascii art in WindingCurrentTarget like you planned.

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp [hunk 7]
```diff
@@ -255,74 +273,100 @@ void ElectroHydraulicPTO::PreUpdate(
...
+  this->dataPtr->x[2] = this->dataPtr->functor.Ri * PBus / this->dataPtr->Ve + this->dataPtr->Ve;
+
+
+  // solver.solveNumericalDiff will compute Jacobian numerically rather than obtain from user
+  const int solver_info = solver.solveNumericalDiff(this->dataPtr->x);
+  if (solver_info != 1) {
+    std::cout << "=================================" << std::endl;
+    std::cout << "Warning: Numericals solver in ElectroHydraulicPTO did not converge" << std::endl;
+    std::cout << "solver info: [" << solver_info << "]" << std::endl;
+    std::cout << "=================================" << std::endl;
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp:314]: perhaps change `std::cout` to `igndbg`?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 8]
```diff
@@ -47,81 +48,192 @@ struct Functor
...
+
+  // Friction loss is characterized in 2022 PTO simulation paper
+  // Friction loss is a function of RPM
+  // Units of power returned
+  double MotorDriveFrictionLoss(double N) const
+  {
+    double tau_c = 0.1;             // N-m
+    double k_v = .06 / 1000;             // N-m/RPM
+    double k_th = 20;
+    return fabs((tau_c * tanh(2 * M_PI * N / 60 / k_th) + k_v * N / 1000) * N);
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:129]: I get nervous about integer literals in division

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 9]
```diff
@@ -47,81 +48,192 @@ struct Functor
...
+      for (int j = 0; j < n; j++) {
+        fjac(k, j) = 0.;
+      }
+      fjac(k, k) = 3. - 4. * x[k];
+      if (k) {fjac(k, k - 1) = -1.;}
+      if (k != n - 1) {fjac(k, k + 1) = -2.;}
+    }
+    return 0;
+  }
+#endif
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:212]: ok to remove this section?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 10]
```diff
@@ -47,81 +48,192 @@ struct Functor
...
+  0.0, 100.0, 200.0, 300.0, 400.0, 500.0, 600.0, 700.0,
+  800.0, 900.0, 1000.0, 1500.0, 2000.0, 2500.0, 3000.0, 3500.0,
+  4000.0, 4500.0, 5000.0, 5500.0, 6000.0, 15000.0};
+
+const std::vector<double> ElectroHydraulicSoln::Eff_M{
+  1.0, 0.950, 0.9460, 0.9450, 0.9440, 0.9430, 0.9420, 0.9410,
+  0.9400, 0.9390, 0.9380, 0.9370, 0.9360, 0.9370, 0.9360, 0.9350,
+  0.9320, 0.9300, 0.9200, 0.9100, 0.9000, 0.8400};
+#endif
+#if 0
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:236]: ok to remove section?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 11]
```diff
@@ -47,81 +48,192 @@ struct Functor
...
+      }
+      fjac(k, k) = 3. - 4. * x[k];
+      if (k) {fjac(k, k - 1) = -1.;}
+      if (k != n - 1) {fjac(k, k + 1) = -2.;}
+    }
+    return 0;
+  }
+#endif
 };
+#if 1
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:214]: ok to make permanent?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 12]
```diff
@@ -132,138 +244,71 @@ const std::vector<double> ElectroHydraulicSoln::Neff{
...
+  double operator()(const double & rpm, const double & pressure) const
+  {
+    if (pressure <= 10.0) {
+      return (2.5871e-14 * rpm + 7.9528e-9) * exp((-3.9e-6 * rpm + 1.25742) * pressure) + 0.1;
+    }
+    return std::max(
+      0.1,
+      (-5.333e-6 * rpm + 0.98) * tanh((-8e-8 * rpm + 0.003) * (pressure + 34.0)));
+  }
 };
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:312]: I think it's ok to remove these as we aren't doing 2D interpolation anymore

File: buoy_gazebo/src/ElectroHydraulicPTO/WindingCurrentTarget.hpp [hunk 13]
```diff
@@ -37,59 +40,76 @@
...
+      J_I *= this->ScaleFactor;
+
+      if (N > 0.0) {
+        J_I *= -this->RetractFactor;
+      }
+    }
+
+    return J_I;
   }
+  */
```
Comment (andermi) [inline buoy_gazebo/src/ElectroHydraulicPTO/WindingCurrentTarget.hpp:104]: delete

File: buoy_tests/launch/eh_solver.launch.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+def find_proc(ld):
+    entities = ld.\
+        _IncludeLaunchDescription__launch_description_source.\
+        _LaunchDescriptionSource__launch_description.\
+        _LaunchDescription__entities
+    for entity in entities:
+        if type(entity) is Node:
+            if type(entity._Action__condition) is UnlessCondition:
+                return entity
```
Comment (andermi) [inline buoy_tests/launch/eh_solver.launch.py:59]: @mjcarroll, I wish there was a better way to do this... (used on line 78)

Comment (mjcarroll) [reply buoy_tests/launch/eh_solver.launch.py:59]: Whoa, what is this trying to do exactly?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp [hunk 15]
```diff
@@ -29,51 +30,55 @@
...
+#include "ElectroHydraulicLoss.hpp"
 
 
 namespace buoy_gazebo
 {
 class ElectroHydraulicPTOPrivate
 {
 public:
-  /// \brief Piston joint entity
+  static constexpr double RPM_TO_RAD_PER_SEC{2.0 * M_PI / 60.0};
```
Comment (mjcarroll) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp:53]: Can we move all of the constants to a common header?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp (hunk 7 above)
Comment (mjcarroll) [reply buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp:314]: I would use `std::stringstream` to append it all together and send it to `igndbg` all at once in that case.

```
std::stringstream ss;
ss << "foo\n";
ss << "bar\n";
igndbg << ss.str();
```

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 16]
```diff
@@ -47,81 +48,192 @@ struct Functor
...
 struct ElectroHydraulicSoln : Functor<double>
 {
 public:
-  const double PressReliefSetPoint = 2850;  // psi
+  const double PressReliefSetPoint = 2850.0;       // psi
   // ~50GPM/600psi ~= .33472 in^3/psi -> Relief valve flow
   // above setpoint, From SUN RPECLAN data sheet
-  const double ReliefValveFlowPerPSI = 50.0 / 600.0;
+  const double ReliefValveFlowPerPSI = 30.0 / 600.0;
   const double CubicInchesPerGallon = 231.0;
```
Comment (mjcarroll) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:88]: More constants that we can put somewhere common?

Comment (hamilton8415) [reply buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:88]: Thanks Michael, I can make those adjustments.  Not entirely sure how high up to put the constants header file, presumably package wide, i'll look to see what makes sense.  Also, I'll leave the "entities" question to andermi

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp (hunk 6 above)
Comment (hamilton8415) [reply buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp:218]: Yeah, that's right.  At the moment I've been leaving that feature disabled as it's a bit second order designed to protect the physical hardware, but it's harder to break metal things in simulation.  It complicates the interpretation of how the PTO model is performing so while I do test those features in "WindingCurrentTarget.hpp", I've been thinking to let things settle in a simpler case before they're put into use...

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp (hunk 16 above)
Comment (mjcarroll) [reply buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:88]: I think `Constants.hpp` in the buoy_utils folder would be reasonable and package wide.

**Comment by hamilton8415** [comment]
Hi All,  I've implemented most of the above suggestions and it's passing CI.  For the constants, as discussed I made a header file in buoy_utils but only put physical constants in there that are useful in a variety of places, unit conversions and similar.  For the constants specific to the ElectroHydraulicPTO, I left those in the class definitions as static constexpr.  There may be some sense in time to pull the constants out that define the exact behavior of this PTO into a header file in ElectroHydraulicPTO, but didn't go for that level of surgery at this time, instead figuring we need to get this merged in.  Thanks for all the help and suggestions,  have a great weekend...

**Comment by andermi** [comment]
LGTM. If @mjcarroll approves, I'll merge.

**Pull request merged**